### PR TITLE
Specify Einsum-7 operation

### PR DESCRIPTION
**Pull request by rkazants**
Ticket: 49323

Signed-off-by: Roman Kazantsev <roman.kazantsev@intel.com>

### CODE REVIEW

File: docs/ops/matrix/Einsum_7.md [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+A = [[[1.0, 2.0, 3.0],
+      [4.0, 5.0, 6.0],
+      [7.0, 8.0, 9.0]]]
+equation = "ijk->kij"
+output = [[[1.0, 4.0, 7.0],
+           [2.0, 5.0, 8.0],
+           [3.0, 6.0, 9.0]]])
+```
+
+Aside a lower case label, ellipsis `...` can be used as a label in a subscript to cover broadcasted dimensions.
```
Comment (mvafin) [inline docs/ops/matrix/Einsum_7.md:116]: Aside a lower case -> In addition to a lower case label, ...

File: docs/ops/matrix/Einsum_7.md [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+The output subscript can be recovered as a sequence of alphabetically sorted labels that do not repeat in the left-hand of the equation.
+For example, `equation = "dbbc,ca"` in implicit mode is the same as `equation = "dbbc,ca->ad"`.
+The equation in implicit mode can set up only subset of Einstein summation conventions so this mode has limitations. For example,
+`equation = "kii->i"` cannot be represented in implicit mode. *Einsum* operation only supports explicit mode. In case of ellipsis label in the left-hand
+of the equation in implicit mode, the ellipsis will come first in the output subscript for the recovery.
+
+The *Einsum* operation has *einsum_path* attribute that can recommend to plugins an order of input contraction. In case of multiple inputs
+the effective order can boost performance and reduce memory costs during *Einsum* computation. The order of applying of *Einsum* operation to inputs 
+makes sense in terms of performance. For example, *Einsum* operation on three inputs (of shapes `[2, 2]`, `[2, 5]` and `[5, 2]`) 
+with `equation="ij,jk,kl->il"` will behave more effective if the operation performs on the second and the third inputs first to avoid
```
Comment (mvafin) [inline docs/ops/matrix/Einsum_7.md:161]: effective -> effectively

Comment (mvafin) [inline docs/ops/matrix/Einsum_7.md:161]: performs -> is performed

File: docs/ops/matrix/Einsum_7.md [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+**Attributes**:
+
+* *equation*
+
+  * **Description**: it defines Einstein summation convention on input operands. The equation must be in explicit mode, i.e. the output subscript must be in place
+  * **Range of values**: the equation format is described above
+  * **Type**: string
+  * **Required**: *yes*
+
+* *einsum_path*
```
Comment (mvafin) [inline docs/ops/matrix/Einsum_7.md:177]: Do we really want this not required attribute specified? Would like to see plugin teams input on this

File: docs/ops/matrix/Einsum_7.md [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+inner (or dot), outer and cross products; transpose; trace and diagonal extraction.
+Also, the single *Einsum* operation can express complex combination of these common linear algebraic tensor operations on multiple operands,
+for example, a dot product of a diagonal extracted from a tensor with shape `[5, 5]` and 5D vector is performed by the single *Einsum* operation.
+The Einstein summation convention on input tensors is defined by `equation` that is a mandatory attribute of *Einsum* operation.
+The format of `equation` is described below.
+
+In explicit mode, the einsum `equation` has the output subscript separated from the input subscripts by `->` and has the following format for `n` operands: 
+`<subscript for input1>, <subscript for input2>, ..., <subscript for inputn> -> <subscript for output>`.
+Each input subscript `<subscript for input1>` contains a sequence of labels (lower case letters `['a',...,'z']`) where each label refers to a dimension of 
+the corresponsing operand. The appereance of labels in a subscript in alphabetical order is not needed. The subscript for a scalar input is empty.
```
Comment (mvafin) [inline docs/ops/matrix/Einsum_7.md:19]: is not needed -> is not required

File: docs/ops/matrix/Einsum_7.md [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+Lastly, the output subscript defines order of output dimensions and which dimensions are sum-reduced.
+Dimensions corresponding to absent labels in the output subscript are sum-reduced. The final result for the considered example is of shape equal to `[3,2]`
+where dimensions with labels `b` and `d` are reduced and the transpose is applied to get output layout `ca`.
+
+**NOTE**: *Einsum* operation can perform on the single operand and in this case the operation can transpose the input and reduce its dimensions.
+
+**NOTE**: Input ranks must be equal to a length of corresponding subscripts and dimensions to which the same labels in input subscripts 
+correspond must be equal in size.
+
+**NOTE**: A labels can repeat in the same input subscript, for example, `equation` equal to `aac,abd,ddde`. In this case the corresponding dimensions
```
Comment (mvafin) [inline docs/ops/matrix/Einsum_7.md:42]: A labels -> "Labels" or "a label"

File: docs/ops/matrix/Einsum_7.md [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+must match in size and the operand the operand will be replaced by its diagonal along these dimensions.
+For example, *Einsum* operation on the single 3D tensor of shape `[2, 4, 5, 4]` with `equation` equal to `ijkj->ij`.
+
+**NOTE**: The specification considers the primitive algorithm for *Einsum* operation for better understanding of the operation
```
Comment (mvafin) [inline docs/ops/matrix/Einsum_7.md:46]: Not sure if this Note is really needed. This is a common-sense that specification is not an implementation recomendation.

File: docs/ops/matrix/Einsum_7.md [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+and does not recommend it for implementation.
+
+**NOTE**: The described algorithm can be improved by immediate dimension sum-reduction in the intermediate results if the corresponding labels absent in
```
Comment (mvafin) [inline docs/ops/matrix/Einsum_7.md:49]: in the intermediate results -> of the intermediate results

File: docs/ops/matrix/Einsum_7.md [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+the input subscripts of subsequent inputs and the output subscript. It can significantly boosts performance and reduces memory costs.
+In the considered example, after the first step we can reduce a dimension corresponding to a label `d`.
+
+The output shape can be easily computed using the output subscript. It needs to concatenate sizes of dimensions to which labels in output subscript
+correspond in the specified order.
```
Comment (mvafin) [inline docs/ops/matrix/Einsum_7.md:54]: It needs to concatenate sizes of dimensions to which labels in output subscript
correspond in the specified order. -> don't like the wording, but hard to come out with a better sentence, could you rephrase it?

**Review by mvafin — CHANGES_REQUESTED** [review]

File: docs/ops/matrix/Einsum_7.md (hunk 2 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:161]: "effective" is also correct here since "more" is used with adjective.

Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:161]: "performs" is correct here. I try to avoid "passive voice" where it is possible. Anyway, I asked our tech-writers and they would check it.

File: docs/ops/matrix/Einsum_7.md (hunk 3 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:177]: Yes, on today's meeting I will raise this question.

File: docs/ops/matrix/Einsum_7.md (hunk 6 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:46]: I think extra reminder will not disturb the specification.

File: docs/ops/matrix/Einsum_7.md (hunk 8 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:54]: I can rephrase these two sentences as follows:
"The output shape is computed by concatenation of dimension sizes to which labels in the output subscript correspond in the specified order."

File: docs/ops/matrix/Einsum_7.md [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+Dimensions corresponding to absent labels in the output subscript are sum-reduced. The final result for the considered example is of shape equal to `[3,2]`
+where dimensions with labels `b` and `d` are reduced and the transpose is applied to get output layout `ca`.
+
+**NOTE**: *Einsum* operation can perform on the single operand and in this case the operation can transpose the input and reduce its dimensions.
+
+**NOTE**: Input ranks must be equal to a length of corresponding subscripts and dimensions to which the same labels in input subscripts 
+correspond must be equal in size.
+
+**NOTE**: A labels can repeat in the same input subscript, for example, `equation` equal to `aac,abd,ddde`. In this case the corresponding dimensions
+must match in size and the operand the operand will be replaced by its diagonal along these dimensions.
```
[the remark this answers is not in the document]
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Done.

File: docs/ops/matrix/Einsum_7.md (hunk 1 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Fixed.

File: docs/ops/matrix/Einsum_7.md (hunk 4 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Done.

File: docs/ops/matrix/Einsum_7.md (hunk 5 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Done

File: docs/ops/matrix/Einsum_7.md (hunk 7 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Fixed.

File: docs/ops/matrix/Einsum_7.md (hunk 8 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Done.

**Review by mvafin — APPROVED** [review]

File: docs/ops/matrix/Einsum_7.md [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+The resulted shape of *Einsum* operation will be `[11, 11, 7, 4]` since a dimension labeled with `a` is left alongside with broadcasted dimensions.
+
+Example 6 shows how *Einsum* operates on the single input with an equation containing ellipsis:
+
+```
+A = [[1.0, 2.0, 3.0],
+     [4.0, 5.0, 6.0],
+     [7.0, 8.0, 9.0]]
+equation = "a...->..."
+output = [12.0, 15.0, 17.0])
```
Comment (rkazants) [inline docs/ops/matrix/Einsum_7.md:132]: `)` - extra. needs to be removed

File: docs/ops/matrix/Einsum_7.md [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+a2 = [4.0, 5.0, 6.0]
+equation = "i,i->"
+output = 32.0
+```
+
+Example 2 shows how *Einsum* computes matrix-vector multiplication:
+
+```
+A = [[1.0, 2.0, 3.0],
+     [1.0, 2.0, 3.0]])
```
Comment (rkazants) [inline docs/ops/matrix/Einsum_7.md:68]: ')' is extra and needs to be removed. The same comment against one below.

File: docs/ops/matrix/Einsum_7.md [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+For example, `equation = "dbbc,ca"` in implicit mode is the same as `equation = "dbbc,ca->ad"`.
+The equation in implicit mode can set up only subset of Einstein summation conventions so this mode has limitations. For example,
+`equation = "kii->i"` cannot be represented in implicit mode. *Einsum* operation only supports explicit mode. In case of ellipsis label in the left-hand
+of the equation in implicit mode, the ellipsis will come first in the output subscript for the recovery.
+
+The *Einsum* operation has *einsum_path* attribute that can recommend to plugins an order of input contraction. In case of multiple inputs
+the effective order can boost performance and reduce memory costs during *Einsum* computation. The order of applying of *Einsum* operation to inputs 
+makes sense in terms of performance. For example, *Einsum* operation on three inputs (of shapes `[2, 2]`, `[2, 5]` and `[5, 2]`) 
+with `equation="ij,jk,kl->il"` will behave more effective if the operation performs on the second and the third inputs first to avoid 
+a larger size dimension, i.e. the dimension which size is equal to `5`. So *einsum_path* can record an order of applying *Einsum* operation.
```
Comment (tsavina) [inline docs/ops/matrix/Einsum_7.md:161]: i.e. ---> that is

File: docs/ops/matrix/Einsum_7.md [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+`equation = "kii->i"` cannot be represented in implicit mode. *Einsum* operation only supports explicit mode. In case of ellipsis label in the left-hand
+of the equation in implicit mode, the ellipsis will come first in the output subscript for the recovery.
+
+The *Einsum* operation has *einsum_path* attribute that can recommend to plugins an order of input contraction. In case of multiple inputs
+the effective order can boost performance and reduce memory costs during *Einsum* computation. The order of applying of *Einsum* operation to inputs 
+makes sense in terms of performance. For example, *Einsum* operation on three inputs (of shapes `[2, 2]`, `[2, 5]` and `[5, 2]`) 
+with `equation="ij,jk,kl->il"` will behave more effective if the operation performs on the second and the third inputs first to avoid 
+a larger size dimension, i.e. the dimension which size is equal to `5`. So *einsum_path* can record an order of applying *Einsum* operation. 
+For this example it will be equal to `"[(1, 2), (0, 1)]"`. The format of *einsum_path* is in string format and represents a list of tuples with operand
```
Comment (tsavina) [inline docs/ops/matrix/Einsum_7.md:162]: For this example it will be equal ---> For this example, it will be equal

File: docs/ops/matrix/Einsum_7.md [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+indices (the start index is `0`). The value of *einsum_path* indicates which inputs of *Einsum* operation should be contracted first,
```
Comment (tsavina) [inline docs/ops/matrix/Einsum_7.md:163]: indices (the start index is `0`) ---> indices, where the start index is `0`

File: docs/ops/matrix/Einsum_7.md [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+
+The *Einsum* operation has *einsum_path* attribute that can recommend to plugins an order of input contraction. In case of multiple inputs
+the effective order can boost performance and reduce memory costs during *Einsum* computation. The order of applying of *Einsum* operation to inputs 
+makes sense in terms of performance. For example, *Einsum* operation on three inputs (of shapes `[2, 2]`, `[2, 5]` and `[5, 2]`) 
+with `equation="ij,jk,kl->il"` will behave more effective if the operation performs on the second and the third inputs first to avoid 
+a larger size dimension, i.e. the dimension which size is equal to `5`. So *einsum_path* can record an order of applying *Einsum* operation. 
+For this example it will be equal to `"[(1, 2), (0, 1)]"`. The format of *einsum_path* is in string format and represents a list of tuples with operand
+indices (the start index is `0`). The value of *einsum_path* indicates which inputs of *Einsum* operation should be contracted first, 
+the result of this contraction is then appended to the end of the contraction list. This list can then be iterated over until
```
Comment (tsavina) [inline docs/ops/matrix/Einsum_7.md:164]: can then be ---> can be

File: docs/ops/matrix/Einsum_7.md [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+For this example it will be equal to `"[(1, 2), (0, 1)]"`. The format of *einsum_path* is in string format and represents a list of tuples with operand
+indices (the start index is `0`). The value of *einsum_path* indicates which inputs of *Einsum* operation should be contracted first, 
+the result of this contraction is then appended to the end of the contraction list. This list can then be iterated over until 
+all intermediate contractions are complete.
+
+**Attributes**:
+
+* *equation*
+
+  * **Description**: it defines Einstein summation convention on input operands. The equation must be in explicit mode, i.e. the output subscript must be in place
```
Comment (tsavina) [inline docs/ops/matrix/Einsum_7.md:171]: i.e. ---> meaning

File: docs/ops/matrix/Einsum_7.md [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+  * **Range of values**: the equation format is described above
+  * **Type**: string
+  * **Required**: *yes*
+
+* *einsum_path*
+
+  * **Description**: it recommends the pre-computed (sub-)optimal path of Einstein summation on multiple inputs. The plugin can ignore it.
```
Comment (tsavina) [inline docs/ops/matrix/Einsum_7.md:178]: pre-computed ---> precomputed

File: docs/ops/matrix/Einsum_7.md [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+**NOTE**: Deep learning frameworks support implicit mode (a classical form of Einstein summation) where the equation does not have the output subscript 
+and has the following format: `<subscript for input1>, <subscript for input2>, ..., <subscript for inputn>`.
+The equation consists of only input subscripts for each operand.
+The output subscript can be recovered as a sequence of alphabetically sorted labels that do not repeat in the left-hand of the equation.
+For example, `equation = "dbbc,ca"` in implicit mode is the same as `equation = "dbbc,ca->ad"`.
+The equation in implicit mode can set up only subset of Einstein summation conventions so this mode has limitations. For example,
+`equation = "kii->i"` cannot be represented in implicit mode. *Einsum* operation only supports explicit mode. In case of ellipsis label in the left-hand
+of the equation in implicit mode, the ellipsis will come first in the output subscript for the recovery.
+
+The *Einsum* operation has *einsum_path* attribute that can recommend to plugins an order of input contraction. In case of multiple inputs
```
Comment (rkazants) [inline docs/ops/matrix/Einsum_7.md:157]: remove `einsum_path` since it is not needed.

[the remark this answers is not in the document]
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:157]: We agreed to remove this attribute and related text at all.

File: docs/ops/matrix/Einsum_7.md [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+the effective order can boost performance and reduce memory costs during *Einsum* computation. The order of applying of *Einsum* operation to inputs 
+makes sense in terms of performance. For example, *Einsum* operation on three inputs (of shapes `[2, 2]`, `[2, 5]` and `[5, 2]`)
```
[the remark this answers is not in the document]
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:159]: We agreed to remove `einsum_path` attribute and related text at all.

File: docs/ops/matrix/Einsum_7.md [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+The output subscript can be recovered as a sequence of alphabetically sorted labels that do not repeat in the left-hand of the equation.
+For example, `equation = "dbbc,ca"` in implicit mode is the same as `equation = "dbbc,ca->ad"`.
+The equation in implicit mode can set up only subset of Einstein summation conventions so this mode has limitations. For example,
+`equation = "kii->i"` cannot be represented in implicit mode. *Einsum* operation only supports explicit mode. In case of ellipsis label in the left-hand
+of the equation in implicit mode, the ellipsis will come first in the output subscript for the recovery.
+
+The *Einsum* operation has *einsum_path* attribute that can recommend to plugins an order of input contraction. In case of multiple inputs
+the effective order can boost performance and reduce memory costs during *Einsum* computation. The order of applying of *Einsum* operation to inputs 
+makes sense in terms of performance. For example, *Einsum* operation on three inputs (of shapes `[2, 2]`, `[2, 5]` and `[5, 2]`) 
+with `equation="ij,jk,kl->il"` will behave more effective if the operation performs on the second and the third inputs first to avoid
```
[the remark this answers is not in the document]
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:160]: We agreed to remove `einsum_path` attribute and related text at all.

File: docs/ops/matrix/Einsum_7.md (hunk 12 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:161]: We agreed to remove `einsum_path` attribute and related text at all.

File: docs/ops/matrix/Einsum_7.md (hunk 13 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:162]: We agreed to remove `einsum_path` attribute and related text at all.

File: docs/ops/matrix/Einsum_7.md (hunk 14 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:163]: We agreed to remove `einsum_path` attribute and related text at all.

File: docs/ops/matrix/Einsum_7.md (hunk 15 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:164]: We agreed to remove `einsum_path` attribute and related text at all.

File: docs/ops/matrix/Einsum_7.md (hunk 17 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:178]: We agreed to remove `einsum_path` attribute and related text at all.

File: docs/ops/matrix/Einsum_7.md [hunk 21]
```diff
@@ -0,0 +1,228 @@
...
+
+**Versioned name**: *Einsum-7*
+
+**Category**: Matrix multiplication
+
+**Short description**: *Einsum* performs the Einstein summation convention on the operands.
+
+**Detailed description**: *Einsum* can represent many common multidimensional linear algebraic tensor operations: matrix multiplication;
+inner (or dot), outer and cross products; transpose; trace and diagonal extraction.
+Also, the single *Einsum* operation can express complex combination of these common linear algebraic tensor operations on multiple operands,
```
Comment (tsavina) [inline docs/ops/matrix/Einsum_7.md:11]: the single *Einsum* operation ---> a single *Einsum* operation

File: docs/ops/matrix/Einsum_7.md [hunk 22]
```diff
@@ -0,0 +1,228 @@
...
+The Einstein summation convention on input tensors is defined by `equation`, which is a mandatory attribute of *Einsum* operation.
+The format of `equation` is described below.
+
+In explicit mode, the einsum `equation` has the output subscript separated from the input subscripts by `->`, and has the following format for `n` operands: 
+`<subscript for input1>, <subscript for input2>, ..., <subscript for inputn> -> <subscript for output>`.
+Each input subscript `<subscript for input1>` contains a sequence of labels (lowercase letters `['a',...,'z']`), where each label refers to a dimension of 
+the corresponsing operand. Labels do not need to appear in a subscript in alphabetical order. The subscript for a scalar input is empty.
+The input subscripts are separated with a comma `,`.
+The output subscript `<subscript for output>` is separated from the input subscripts by `->` and represents a sequence of labels
+(lowercase letters `['a',...,'z']`). A length of the input subscript matches a rank of the input. The input subscript is empty for a scalar input.
```
[the remark this answers is not in the document]
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Fixed a bit different as agreed.

File: docs/ops/matrix/Einsum_7.md [hunk 23]
```diff
@@ -0,0 +1,228 @@
...
+For example, `equation = "dbbc,ca"` in implicit mode is the same as `equation = "dbbc,ca->ad"`.
+The equation in implicit mode can set up only subset of Einstein summation conventions, so this mode has limitations. For example,
+`equation = "kii->i"` cannot be represented in implicit mode. *Einsum* operation only supports explicit mode.
+In case ellipsis label is in the left-hand side of the equation in implicit mode, the ellipsis comes first in the output subscript for the recovery.
+
+**Attributes**:
+
+* *equation*
+
+  * **Description**: it defines Einstein summation convention on input operands. The equation must be in explicit mode meaning the output subscript must be in place
```
[the remark this answers is not in the document]
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Fixed.

File: docs/ops/matrix/Einsum_7.md (hunk 21 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Fixed.

**Review by tsavina — APPROVED** [review]

File: docs/ops/matrix/Einsum_7.md [hunk 24]
```diff
@@ -0,0 +1,228 @@
...
+**Versioned name**: *Einsum-7*
+
+**Category**: Matrix multiplication
+
+**Short description**: *Einsum* performs the Einstein summation convention on the operands.
+
+**Detailed description**: *Einsum* can represent many common multidimensional linear algebraic tensor operations: matrix multiplication;
+inner (or dot), outer and cross products; transpose; trace and diagonal extraction.
+Also, a single *Einsum* operation can express complex combination of these common linear algebraic tensor operations on multiple operands,
+for example, a dot product of a diagonal, extracted from a tensor with shape [5, 5], and 5D vector is performed by single Einsum operation.
```
Comment (lazarevevgeny) [inline docs/ops/matrix/Einsum_7.md:12]: nit: wrap "[5, 5]" with back ticks "`"

File: docs/ops/matrix/Einsum_7.md [hunk 25]
```diff
@@ -0,0 +1,228 @@
...
+The Einstein summation convention on input tensors is defined by `equation`, which is a mandatory attribute of *Einsum* operation.
+The format of `equation` is described below.
+
+In explicit mode, the einsum `equation` has the output subscript separated from the input subscripts by `->`, and has the following format for `n` operands: 
+`<subscript for input1>, <subscript for input2>, ..., <subscript for inputn> -> <subscript for output>`.
+Each input subscript `<subscript for input1>` contains a sequence of labels (lowercase letters `['a',...,'z']`), where each label refers to a dimension of 
+the corresponsing operand. Labels do not need to appear in a subscript in alphabetical order. The subscript for a scalar input is empty.
+The input subscripts are separated with a comma `,`.
+The output subscript `<subscript for output>` is separated from the input subscripts by `->` and represents a sequence of labels
```
Comment (lazarevevgeny) [inline docs/ops/matrix/Einsum_7.md:21]: nit: we already said that the output subscript is separated from the input by '->' (in the beginning of the paragraph), so probably here we may omit this and just write: "The output subscript `<subscript for output>` represents a sequence of labels..."

File: docs/ops/matrix/Einsum_7.md [hunk 26]
```diff
@@ -0,0 +1,228 @@
...
+(lowercase letters `['a',...,'z']`). The length of an input subscript matches a rank of the input. The input subscript is empty for a scalar input.
+
+*Einsum* operation on multiple inputs can be treated as several consecutive *Einsum* operations. In the first step, *Einsum* applies the first two inputs. 
+In the second step, it operates on the result of the first step and the third input, and so forth.
+*Einsum* operates on two operands similar to element-wise multiplication by all pairs of batches from both operands. The batch dimensions are defined
+with labels that enter only one subscript from the two corresponding subscripts.
```
Comment (lazarevevgeny) [inline docs/ops/matrix/Einsum_7.md:27]: May be change "The batch dimensions are defined with labels that enter only one subscript from the two corresponding subscripts." to "The batch dimensions are defined with labels belonging to only one of the two input subscripts."

File: docs/ops/matrix/Einsum_7.md [hunk 27]
```diff
@@ -0,0 +1,228 @@
...
+output = [[[1.0, 4.0, 7.0],
+           [2.0, 5.0, 8.0],
+           [3.0, 6.0, 9.0]]]
+```
+
+In addition to a lowercase label, ellipsis `...` can be used as a label in a subscript to cover broadcasted dimensions.
+Each input subscript can contain at most one ellipsis. For example, the ellipsis in input subscript `a...bc` for five rank tensor covers 
+the second and third dimensions. In case input subscripts contain ellipsis for several operands, the dimensions covered by the ellipsis
+must be broadcastable to satisfy numpy broadcasting (or multidirectional broadcasting) rules available in
+<a href="https://numpy.org/doc/stable/user/basics.broadcasting.html#general-broadcasting-rules">numpy broadcasting documentation</a>.
```
Comment (lazarevevgeny) [inline docs/ops/matrix/Einsum_7.md:118]: we have our own document describing broadcasting rules `docs/ops/broadcast_rules.md`

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: docs/ops/matrix/Einsum_7.md (hunk 24 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Fixed.

File: docs/ops/matrix/Einsum_7.md (hunk 25 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Fixed.

File: docs/ops/matrix/Einsum_7.md (hunk 26 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Fixed.

File: docs/ops/matrix/Einsum_7.md (hunk 27 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Fixed.

File: docs/ops/matrix/Einsum_7.md [hunk 28]
```diff
@@ -0,0 +1,228 @@
...
+```
+
+**NOTE**: Deep learning frameworks support implicit mode (a classical form of Einstein summation) where the equation does not have the output subscript 
+and has the following format: `<subscript for input1>, <subscript for input2>, ..., <subscript for inputn>`.
+The equation consists of only input subscripts for each operand.
+The output subscript can be recovered as a sequence of alphabetically sorted labels that are not repeated in the left-hand side of the equation.
+For example, `equation = "dbbc,ca"` in implicit mode is the same as `equation = "dbbc,ca->ad"`.
+The equation in implicit mode can set up only subset of Einstein summation conventions, so this mode has limitations. For example,
+`equation = "kii->i"` cannot be represented in implicit mode. *Einsum* operation only supports explicit mode.
+In case ellipsis label is in the left-hand side of the equation in implicit mode, the ellipsis comes first in the output subscript for the recovery.
```
Comment (lazarevevgeny) [inline docs/ops/matrix/Einsum_7.md:154]: Why do we limit our customers to only explicit mode? We can easily convert the implicit mode to explicit mode internally for easier handling by plugin, but from the API perspective I vote for allowing implicit mode.

File: docs/ops/matrix/Einsum_7.md [hunk 29]
```diff
@@ -0,0 +1,228 @@
...
+          [[3.0, 6.0, 9.0]]]
+```
+
+In addition to a lowercase label, ellipsis `...` can be used as a label in a subscript to cover broadcasted dimensions.
+Each input subscript can contain at most one ellipsis. For example, the ellipsis in input subscript `a...bc` for five rank tensor covers 
+the second and third dimensions. In case input subscripts contain ellipsis for several operands, the dimensions covered by the ellipsis
+must be broadcastable to satisfy numpy broadcasting (or multidirectional broadcasting) rules available in
+[Broadcast Rules For Elementwise Operations](../broadcast_rules.md).
+If at least one input subscript contains an ellipsis, the output subscript must always contain one ellipsis.
+For example, *Einsum* operation on two inputs of shapes `[11, 1, 4, 3]` and `[3, 11, 7, 1]` with `equation="a...b,b...->a..."`
```
Comment (lazarevevgeny) [inline docs/ops/matrix/Einsum_7.md:120]: please, change the value 11 in one of the input tensors to some unique number, for example, 9

File: docs/ops/matrix/Einsum_7.md [hunk 30]
```diff
@@ -0,0 +1,228 @@
...
+        </port>
+        <port id="3">
+            <dim>2</dim>
+            <dim>4</dim>
+            <dim>7</dim>
+        </port>
+    </input>
+    <output>
+        <port id="4">
+            <dim>4</dim>
```
Comment (lazarevevgeny) [inline docs/ops/matrix/Einsum_7.md:222]: I didn't get why this dimension is equal to 4. Can you explain the evaluation step by step?

File: docs/ops/matrix/Einsum_7.md (hunk 28 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:154]: Implicit mode can be easily represented with explicit mode but not vise versa. We can offload recovering output subscript into importer, MO and different other bridge.

Comment (lazarevevgeny) [reply docs/ops/matrix/Einsum_7.md:154]: What if customer builds the model using C++ or Python API? In this case there is no any "importer" in between and they are forced to use explicit mode.

Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:154]: In any case the customer has to pay attention to that our opset is different from frameworks opset. I think it is a drawback not to have the output subscript that I am going to fix in our opset because the implicit mode for all frameworks does not declare definite order of dimensions.

File: docs/ops/matrix/Einsum_7.md (hunk 30 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:222]: This logic was explained above in paragraph about ellipsis. In this case, we have two operands with broadcasted dimensions: the first operand with broadcasted dimension ([4]) and the second operand with broadcasted dimension ([1]). They are broadcasted to the common shape [4]

File: docs/ops/matrix/Einsum_7.md [hunk 31]
```diff
@@ -0,0 +1,228 @@
...
+**NOTE**: A label can be repeated in the same input subscript, for example, `equation` equal to `aac,abd,ddde`. In this case, the corresponding dimensions
+must match in size, and the operand is replaced by its diagonal along these dimensions.
+For example, *Einsum* operation on the single 3D tensor of shape `[2, 4, 5, 4]` with `equation` equal to `ijkj->ij`.
+
+**NOTE**: The specification considers the primitive algorithm for *Einsum* operation for better understanding of the operation
+and does not recommend it for implementation.
+
+**NOTE**: The described algorithm can be improved by immediate dimension sum-reduction of the intermediate results if the corresponding labels are absent 
+in the input subscripts of subsequent inputs and the output subscript. It can significantly boost performance and reduce memory costs.
+In the considered example, after the first step you can reduce the dimension corresponding to the label `d`.
```
[the remark this answers is not in the document]
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md:53]: Guys, let us stay these notes. They are not redundant to have the common idea about this operation and a bit more.
They reflect a sort of respect to readers.

File: docs/ops/matrix/Einsum_7.md [hunk 32]
```diff
@@ -0,0 +1,228 @@
...
+The resulted shape of *Einsum* operation will be `[11, 11, 7, 4]` since the dimension labeled with `a` is left with broadcasted dimensions.
+
+Example 6 shows how *Einsum* operates on the single input with an equation containing ellipsis:
+
+```
+A = [[1.0, 2.0, 3.0],
+     [4.0, 5.0, 6.0],
+     [7.0, 8.0, 9.0]]
+equation = "a...->..."
+output = [12.0, 15.0, 17.0]
```
[the remark this answers is not in the document]
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Great catch. Done.

File: docs/ops/matrix/Einsum_7.md (hunk 28 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: We agreed to support implicit mode and include capital letters in equation. The text is changed based on this comment.

File: docs/ops/matrix/Einsum_7.md (hunk 29 above)
Comment (rkazants) [reply docs/ops/matrix/Einsum_7.md]: Fixed.

**Review by iimironov — APPROVED** [review]

**Review by lazarevevgeny — APPROVED** [review]
